691 - Add tapestry setting to load a map of earth instead of tapestry

wynnset opened the pull request:
## Changes
* Adds a "Geography Map" option that, when checked, shows a map of earth instead of the tapestry
* Map displays markers for each node and clicking on each marker will show a popup to view the node 
* When this options is checked and the user is logged in:
  * Show a list of nodes on the side for authors to add, edit, and place nodes on the map
  * Show a "Geography" tab on the node modal where the user can enter coordinates for that node
## Screenshots
![image](https://user-images.githubusercontent.com/8981258/99889420-c6688b00-2c09-11eb-8f70-516b73291326.png)
![image](https://user-images.githubusercontent.com/8981258/99889432-f9128380-2c09-11eb-85bc-62a8bc1713db.png)
![image](https://user-images.githubusercontent.com/8981258/99889455-50185880-2c0a-11eb-9d66-e6871e55f021.png)
## Issue Linkage
Closes #691 
## PR Dependency
Depends on: N/A
## Automated Testing
...

wynnset commented:
@yungalyx testing

wynnset commented:
@yungalyx testing 2

wynnset commented:
@yungalyx getting this error which is preventing the page from loading: 
```
ERROR in ./node_modules/leaflet/dist/leaflet.css 3:0
Module parse failed: Unexpected token (3:0)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
| /* required styles */
|
> .leaflet-pane,
| .leaflet-tile,
| .leaflet-marker-icon,
 @ ./src/components/TapestryApp.vue?vue&type=script&lang=js& (./node_modules/babel-loader/lib!./node_modules/vue-loader/lib??vue-loader-options!./src/components/TapestryApp.vue?vue&type=script&lang=js&) 77:0-34
 @ ./src/components/TapestryApp.vue?vue&type=script&lang=js&
 @ ./src/components/TapestryApp.vue
 @ ./node_modules/babel-loader/lib!./node_modules/vue-loader/lib??vue-loader-options!./src/App.vue?vue&type=script&lang=js&
 @ ./src/App.vue?vue&type=script&lang=js&
 @ ./src/App.vue
 @ ./src/main.js
 @ multi @babel/polyfill ./src/main.js
```

yungalyx commented:
@wynnset could you try running 'npm install && npm run dev'?

wynnset commented:
@yungalyx that's what I did.

wynnset commented on templates/vue/src/components/TapestryMap.vue:
```diff
@@ -0,0 +1,89 @@
+    <div style="height: 150px overflow: auto;">
+      <p>First marker is placed at {{ withPopup.lat }}, {{ withPopup.lng }}</p>
+      <p> The bounds are: {{ mapBounds }} </p>
+      <p> Settings are: {{this.settings}} </p>
```
You never use `.this` in the template. The template is bound to `this` already so the variables you use in it are assumed to be properties of `this` already. So this should be just `settings`.

wynnset commented on templates/vue/src/components/TapestryMap.vue:
```diff
@@ -0,0 +1,89 @@
+</template>
+
+<script>
+import { latLng, latLngBounds } from "leaflet";
+import { LMap, LTileLayer, LMarker, LPopup, LTooltip } from "vue2-leaflet";
+import { mapState } from 'vuex';
+
+export default {
+  name: "Tapestry Map",
+  components: {
+    LMap,
+    LTileLayer,
+    LMarker,
+    LPopup,
+    LTooltip
+  },
```
If you are not using `LMarker`, `LPopup`, and `LTooltip`, do not include them here (or comment them out for your draft PR). That way, linter will not complain

wynnset commented on templates/vue/src/components/TapestryMap.vue:
```diff
@@ -0,0 +1,89 @@
+</template>
+
+<script>
+import { latLng, latLngBounds } from "leaflet";
```
This would usually get caught by the linter, but in our code, we drop the semicolon at the end of each line.

wynnset commented on templates/vue/src/components/TapestryMap.vue:
```diff
@@ -0,0 +1,89 @@
+    }
+  },
+  methods: {
+    zoomUpdate(zoom) {
+      this.currentZoom = zoom;
+    },
+    centerUpdate(center) {
+      this.currentCenter = center;
+    },
```
Please name your methods with the verb at the beginning, then the subject. So these would be `updateZoom` and `updateCenter`.

wynnset commented on templates/vue/src/components/TapestryMap.vue:
```diff
@@ -0,0 +1,89 @@
+      style="height: 80%"
+      @update:center="centerUpdate"
+      @update:zoom="zoomUpdate"
+     
```
This would also get caught by the linter usually, but there is no need for this empty line here.

wynnset commented on templates/vue/src/components/TapestryMap.vue:
```diff
@@ -0,0 +1,89 @@
+  },
+  data() {
+    return {
+      zoom: 14,
+      center: latLng(49.262796, -123.252165),
+      url: 'https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png',
+      attribution:
+        '&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors',         
+      withPopup: latLng(47.41322, -1.219482),
+      mapBounds: latLngBounds([
+        [90, 180],
+        [-90, -180]
+      ]),
+      withTooltip: latLng(47.41422, -1.250482),
+      showParagraph: false,
+      mapOptions: {
+        zoomSnap: 0.5
+      },
```
You have a few data elements that are not used. Make sure you review and remove them before finalizing your PR. If you think we might want to add/utilize some of them in the future, you can comment them out but if you do comment something out, leave a comment above it to say why it's commented out and what it could be used for.

wynnset reviewed (commented):
@yungalyx I left you some comments here to give you a few tips about coding standards and Vue in general. I know you are still working on this, but thought this might help.

wynnset commented:
@narendrasss might be able to help you out with the css include issue I reported up there - I think it might have something to do with webpack.

yungalyx commented:
@wynnset fixed linting bug, renamed variables (hopefully they're a bit better 😄), and improved the UX of adding map bounds in settings with form validation. Check it out!

yungalyx commented:
@wynnset can you try running this branch when you have time? css issues should be fixed, thanks!

wynnset commented on templates/vue/src/components/SettingsModal.vue:
```diff
@@ -192,11 +247,19 @@ export default {
   computed: {
     ...mapGetters(["tapestryJson"]),
     ...mapState(["settings", "rootId"]),
+    latbound() {
+      return (this.mapBounds.neLat || 90) > (this.mapBounds.swLat || -90)
+    },
+    lngbound() {
+      return (this.mapBounds.neLng || 180) > (this.mapBounds.swLng || -180)
+    },
```
`latbound` and `lngbound` have 2 problems in their naming:
1. `lat` and `bound` are 2 different words and you should be using camelCase, as previously mentioned.
2. `latbound` is not explaining exactly what this variable does. You should name it something that is easy to understand like `latitudeRangeInvalid` or something like that. This will help other devs understand better what that variable is without having to look at its implementation.

wynnset commented on templates/vue/src/components/TapestryApp.vue:
```diff
@@ -50,25 +50,31 @@
 import LockedTooltip from "@/components/LockedTooltip"
 import TapestryFilter from "@/components/TapestryFilter"
 import Helpers from "@/utils/Helpers"
+import "leaflet/dist/leaflet.css"
```
I don't know if it's a good idea to load the leaflet css file here. What if we don't need it? That means it will load all the time even if we don't need it? Maybe a solutions like what's suggested [here](https://stackoverflow.com/questions/52767025/how-to-load-css-file-dynamically) would work. @narendrasss what do you think?

wynnset reviewed (commented):
Good start. The map looks fine now and is loading for me. A few things: 
- I can't seem to be able to pan the map. It just bounces back every time I drag to move the map. 
- I think we should disable scroll to zoom.
- I think the coordinates that are being accepted are for north west and south east, even though you have labelled them as north east / south west. When I enter coordinates for north east and south west, it gives me an error and I have to change them to north west and south east.
Once those are fixed, the next step is to implement actually putting the nodes on the map as discussed. Thanks!

yungalyx commented:
@wynnset i moved the import css import statement from tapestryApp to tapestryMap.. would that maybe improve import so we only import if `<tapestry-map/>` is rendered? 

<code>In TapestryApp.vue:</code> 
`<div v-if="settings.renderMap" id="map-container"> 
      <tapestry-map />
    </div>`

wynnset commented:
Hi @yungalyx I don't think you addressed the items I mentioned in [my previous comment above](https://github.com/wynnset/tapestry-wp/pull/692#pullrequestreview-510152862) yet.

yungalyx commented:
the bounds won't let you pan beyond the coordinates! i added a red box to visually indicate that at the moment.
Scroll has been disabled and bounds will automatically dictate initial zoom level. 
<img width="532" alt="map bound" src="https://user-images.githubusercontent.com/52767711/96630124-6a88aa80-12c9-11eb-9e8a-b5106c6d8c76.png">
<img width="931" alt="Screenshot 2020-10-20 114557" src="https://user-images.githubusercontent.com/52767711/96630424-dcf98a80-12c9-11eb-9309-e426aada786c.png">

wynnset commented:
@yungalyx 
- [ ] The user should still be able to pinch to zoom in/out
- [ ] The zoom icons (+/-) have disappeared since you moved the css import
- [x] Scroll to zoom is no longer working which is great
- [ ] I'm not sure what the purposes of the red line is... We should fix that issue instead. Is there no setting in leaflet to change that? It seems like a very weird behaviour. Maybe you are approaching it the wrong way? We could have a call to try and resolve this together if you want.
- [ ] I still cannot map nodes

yungalyx commented:
Still a few things to incorporate from my point of view:

- [ ] map nodes shouldn't show based on permission settings

- [ ] changes to node map coordinates are not rendered into map unless we refresh the page (i populate the marker array on create ) i.e, no 'live' updates

jaskwong commented:
- when dragging to move the map, this activates drag select 
- it seems that only text nodes can be mapped (for video, h5p, external link, etc. it doesn't show the option to specify coordinates after checking the "show this node on map" box)
- at the ends of the map (ex. at the very top), the map can still be dragged but this just shows the grey background behind it, maybe it would be better at the map bounds to not let it be dragged any further (this is just my opinion, but currently the map can basically be dragged out of view)
- based on the asana task, "we will hide the tapestry of nodes/links...simply list the nodes on the side of the map" so this is probably something to incorporate

I know this is still a draft pr so I apologize in advance if any of this is not currently applicable

narendrasss commented on templates/vue/src/components/TapestryMap.vue:
```diff
@@ -0,0 +1,117 @@
+      })
+    },
+    resetMarkers() {
+      this.markers = []
+      for (const [key, value] of Object.entries(this.nodes)) {
+        if (this.hasMapCoordinates(value.mapCoordinates)) {
+          this.markers.push({
+            id: key,
+            pos: latLng(value.mapCoordinates.lat, value.mapCoordinates.lng),
+          })
+        }
+      }
```
With regards to your message on slack, I would maybe try this:

```suggestion
      const markers = []
      for (const [key, value] of Object.entries(this.nodes)) {
        if (this.hasMapCoordinates(value.mapCoordinates)) {
          markers.push({
            id: key,
            pos: latLng(value.mapCoordinates.lat, value.mapCoordinates.lng),
          })
        }
      }
      this.markers = markers
```

More than that, I would suggest changing markers from a data property to a computed property instead. Using a computed property will tell Vue to keep `markers` in sync with `this.nodes`, which means Vue knows to recalculate when `this.nodes` changes. This removes the duplication of code in `resetMarkers` and `created`, and removes the need for the event bus (which I thought was removed?)

yungalyx commented:
@jaskwong yeah i'm getting an error with the "show this node on map" error too. I'm not sure why because sometimes when I refresh the page, it'll work again..

jaskwong commented:
@yungalyx do you want me to take a look at it?

yungalyx commented:
@jaskwong sure! that would be super helpful, thank you!

jaskwong commented:
@yungalyx I think it's cause for new nodes, node.mapCoordinate isn't actually defined and in CoordinateForm.vue when created() tries to access the lat/lng coordinates it produces an error and same with the b-form-input which stops it from displaying, are you setting node.mapCoordinates for nodes, other than in the initial loading of the tapestry?

A tip for debugging: try opening the console in your browser inspect tool, errors are usually logged there so you can see what they say!

wynnset commented on templates/vue/src/components/NodeModal.vue:
```diff
@@ -257,6 +268,9 @@ export default {
         ? true
         : wpData.wpCanEditTapestry !== ""
     },
+    showMap() {
+      return this.settings.renderMap
+    },
```
I think it would be better if you put `settings.renderMap` directly in your `v-if`. No reason to make the reader go searching for where `showMap` comes from.

wynnset commented on templates/vue/src/components/SettingsModal.vue:
```diff
@@ -109,6 +109,45 @@
           </b-form-group>
+          <b-form-group
+            label="World map"
+            description="Replace Tapestry with a new authoring experience over a map of the Earth"
```
This should read "Replace Tapestry with a map of Earth with placeholders for each node."

wynnset commented on templates/vue/src/components/SettingsModal.vue:
```diff
@@ -109,6 +109,45 @@
+            </b-form-checkbox>
+          </b-form-group>
+          <div v-if="renderMap">
+            <p v-if="!latitudeRangeValid" class="error-message">
+              North coordinate must be greater than south coordinate
+            </p>
+            <p v-if="!longitudeRangeValid" class="error-message">
+              East coordinate must be greater than west coordinate
+            </p>
```
As mentioned before (but not yet done) please use Bootstrap-Vue provided functionality for showing errors. See https://bootstrap-vue.org/docs/components/form-input#contextual-states

wynnset commented on templates/vue/src/components/SettingsModal.vue:
```diff
@@ -109,6 +109,45 @@
+              <b-col sm="3">
+                <span>Northeast bound:</span>
+              </b-col>
+              <b-col sm="3">
+                <b-form-input v-model="mapBounds.neLat" placeholder="90" />
+              </b-col>
+              <b-col sm="3">
+                <b-form-input v-model="mapBounds.neLng" placeholder="180" />
+              </b-col>
```
Please label which one is latitude and which one is longitude.

wynnset commented on templates/vue/src/components/SettingsModal.vue:
```diff
@@ -109,6 +109,45 @@
+                <span>Southwest bound:</span>
+              </b-col>
+              <b-col sm="3">
+                <b-form-input v-model="mapBounds.swLat" placeholder="-90" />
+              </b-col>
+              <b-col sm="3">
+                <b-form-input v-model="mapBounds.swLng" placeholder="-180" />
+              </b-col>
```
Please label which one is latitude and which one is longitude.

wynnset commented on templates/vue/src/components/SettingsModal.vue:
```diff
@@ -213,12 +255,21 @@ export default {
     ...mapGetters(["tapestryJson"]),
     ...mapState(["settings", "rootId"]),
+    // the + before this.mapBounds changes type from string to int for correct comparison
+    latitudeRangeValid() {
+      return (+this.mapBounds.neLat || 90) > (+this.mapBounds.swLat || -90)
+    },
+    longitudeRangeValid() {
+      return (+this.mapBounds.neLng || 180) > (+this.mapBounds.swLng || -180)
+    },
```
You probably also want to check that latitude and longitude are valid values / ranges. I entered `1000` and it accepted it!

wynnset commented on templates/vue/src/components/TapestryMap.vue:
```diff
@@ -0,0 +1,126 @@
+          v-for="marker in markerlocations"
+          :key="marker.id"
+          :lat-lng="marker.pos"
+          :icon="marker.access ? icon : inaccessibleIcon"
```
Can you also add a separate icon for draft nodes? It would be nice if it follows the look and feel of draft nodes (i.e. dashed outline)

wynnset commented on templates/vue/src/components/TapestryMap.vue:
```diff
@@ -0,0 +1,126 @@
+    },
+    markerlocations() {
+      const markers = []
+      for (const [key, value] of Object.entries(this.nodes)) {
```
I suggest naming your variables more clearly like `const [nodeId, node]`.

wynnset commented on templates/vue/src/components/TapestryMap.vue:
```diff
@@ -0,0 +1,126 @@
+<script>
+import "leaflet/dist/leaflet.css"
+import mapMarker from "@/assets/map-marker.png"
+import invalidMarker from "@/assets/map-marker-inaccessible.png"
```
Could you change this icon and put a lock in it instead of the dot?

wynnset commented on templates/vue/src/components/TapestryMap.vue:
```diff
@@ -0,0 +1,126 @@
+        >
+          <l-popup>
+            <div v-if="marker.access" @click="openNode(marker.id)">
+              {{ marker.title }}
```
Can you add a style for this and set `cursor: pointer;` so it appears more like a link? Maybe also make it bold. (This does not apply for the inaccessible nodes)

wynnset commented on templates/vue/src/components/node-modal/CoordinateForm.vue:
```diff
@@ -0,0 +1,77 @@
+        <b-form-input
+          v-model="node.mapCoordinates.lat"
+          :number="true"
+          placeholder="enter latitude coordinate here"
```
Don't forget to always capitalize. You could also shorten it a bit like "Enter latitude".

wynnset commented on templates/vue/src/components/node-modal/CoordinateForm.vue:
```diff
@@ -0,0 +1,77 @@
+        <b-form-input
+          v-model="node.mapCoordinates.lng"
+          :number="true"
+          placeholder="enter longitude coordinate here"
```
Don't forget to always capitalize. You could also shorten it a bit like "Enter longitude".

wynnset commented on templates/vue/src/components/node-modal/CoordinateForm.vue:
```diff
@@ -0,0 +1,77 @@
+    ...mapGetters(["getNode"]),
+  },
+  watch: {
+    /*
+    nodeLat(){
+      this.node.mapCoordinates.lat = this.nodeLat
+    },
+    nodeLng(){
+      this.node.mapCoordinates.lng = this.nodeLng
+    },
+    */
```
Might want to remove this if unused

wynnset commented:
Hi @yungalyx we are almost here! I left you some code comments, but also noticed you have not addressed some comments from @jaskwong above, so I'm assigning back to you. Should be easy fixes. Please check off or mark as resolved when you address each point so we can keep track. Thanks!

wynnset commented on templates/vue/src/components/NodeModal.vue:
```diff
@@ -97,6 +98,14 @@
         </b-tab>
+        <b-tab
+          v-if="showMap"
+          title="Coordinate"
```
Please change the tab to be called "Map". Please also change the variable names to be plural i.e. `coordinates` and the component to `coordinates-form`.

yungalyx commented:
@jaskwong Sean's 809 PR fixes the dragSelect issue on my map! I will start the doing the nodes on the side of map tomorrow, thanks again for the help

yungalyx commented:
@wynnset everything should be fixed, i've included a list to add nodes alongside the map. Only no icons for draft nodes yet because I couldn't find any good ones that were free. All the nice looking ones need a license :0

wynnset commented:
Thanks @yungalyx. The nodes should not appear if the user cannot edit the settings for the tapestry. I'm working on the placeholder icons for you and will update soon.

narendrasss commented on templates/vue/src/components/TapestryMap.vue:
```diff
@@ -0,0 +1,324 @@
+  },
+  data() {
+    return {
+      url: "https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",
+      attribution:
+        '&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors',
+      mapOptions: {
+        zoomSnap: 0.1,
+        scrollWheelZoom: false,
+      },
+      icon: icon({
+        iconUrl: mapMarker,
+        iconSize: [32, 33],
+        iconAnchor: [16, 32],
+        popupAnchor: [0, -20],
+      }),
+      inaccessibleIcon: icon({
+        iconUrl: invalidMarker,
+        iconSize: [32, 33],
+        iconAnchor: [16, 32],
+        popupAnchor: [0, -20],
+      }),
+      draftIcon: icon({
+        iconUrl: draftMarker,
+        iconSize: [32, 33],
+        iconAnchor: [16, 32],
+        popupAnchor: [0, -20],
+      }),
+      worldBounds: latLngBounds([
+        [-90, -180],
+        [90, 180],
+      ]),
```
Do these attributes change? Typically `data` attributes are reserved for those that change throughout the component lifecycle. Otherwise you would put them under `computed`. Part of the reason is performance, but it also shows people who are reading the code that it won't change.

narendrasss commented on templates/vue/src/components/TapestryMap.vue:
```diff
@@ -0,0 +1,324 @@
+</style>
+
+<style lang="scss">
+#map-container .nodes-list > .card .card-body table tbody > tr {
```
This selector looks very specific - not super clear to me what element(s) this selector is actually affecting. If you're overriding bootstrap values, could a [scoped slot](https://bootstrap-vue.org/docs/components/table#custom-data-rendering) work better for this use case?

wynnset replied:
There's only one table element in there. Not sure if the scoped slot can do what is intended as we are using the `stacked` table. This table doesn't provide much in terms of built-in modifiers, so it's necessary to make these minor modifications here to match the intended styling.

narendrasss commented on templates/vue/src/components/TapestryMap.vue:
```diff
@@ -0,0 +1,324 @@
+    selectedId() {
+      return this.$route.params.nodeId
+    },
+    isLoggedIn() {
+      return isLoggedIn
+    },
+    canEdit() {
+      return wpData.wpCanEditTapestry === "1"
+    },
```
Can you change this to use the [wp module](https://github.com/wynnset/tapestry-wp/blob/master/templates/vue/src/services/wp.js) instead for consistency?

narendrasss commented on templates/vue/src/components/TapestryMap.vue:
```diff
@@ -0,0 +1,324 @@
+import { mapState } from "vuex"
+import { names } from "@/config/routes"
+import Helpers from "@/utils/Helpers"
+import { isLoggedIn } from "@/utils/wp"
```
This import may not be necessary after the change to the `wp` module.

narendrasss reviewed (changes requested):
A few nitpicks, but otherwise looks good!

wynnset commented:
@yungalyx could you please add automated tests to this?

rlnsy commented:
@yungalyx @wynnset This works for me! Anything in particular I should test? I succesfully created a new video Node and placed it on the map. Also, looking at the merge conflicts they appear to just be indentation related?

yungalyx commented:
@rlnsy awesome! a lot of manual testing was done so I'm glad it works well. Only concern at the moment is how much automated tests need to be written before merge. thoughts? @wynnset

narendrasss commented:
@yungalyx ideally everything that you're manually testing should be automated. I also like to add in edge cases in the form of integration tests whenever that's applicable

wynnset commented on templates/vue/debug.log:
```diff
@@ -0,0 +1,4 @@
+[1111/155015.365:ERROR:directory_reader_win.cc(43)] FindFirstFile: The system cannot find the path specified. (0x3)
+[1112/111843.045:ERROR:directory_reader_win.cc(43)] FindFirstFile: The system cannot find the path specified. (0x3)
+[1113/123327.742:ERROR:directory_reader_win.cc(43)] FindFirstFile: The system cannot find the path specified. (0x3)
+[1113/140254.460:ERROR:directory_reader_win.cc(43)] FindFirstFile: The system cannot find the path specified. (0x3)
```
@yungalyx what is this file and why is it committed? I generally avoid using `git commit -a` because of this.

wynnset commented:
> @yungalyx ideally everything that you're manually testing should be automated. I also like to add in edge cases in the form of integration tests whenever that's applicable

I agree with @narendrasss - here are a few more things that I think could be added (feel free to add more if you think of some):
- [ ] Ensure changing map coordinate settings change the coordinates
- [ ] Ensure adding a node to map works
- [ ] Ensure deleting a node from map works
- [ ] Ensure editing the coordinates for a node updates its position on the map
- [ ] Ensure clicking on a pop-up link opens the node
